pkg/ottl/parser_collection.go

```diff
@@ -0,0 +1,289 @@
+		}
+		mp.contextParsers[context] = &parserCollectionParser[S]{
+			ottlParser:          newParserWrapper[K, S](parser),
+			statementsConverter: newStatementsConverterWrapper(converter),
+		}
+		return nil
+	}
+}
+
+// WithParserCollectionErrorMode has no effect on the ParserCollection, but might be used
+// by the ParsedStatementConverter functions to handle/create StatementSequence.
+func WithParserCollectionErrorMode[S StatementsGetter, R any](errorMode ErrorMode) ParserCollectionOption[S, R] {
+	return func(tp *ParserCollection[S, R]) error {
+		tp.ErrorMode = errorMode
+		return nil
+	}
+}
+
+// EnableParserCollectionModifiedStatementLogging controls the statements modification logs.
+// When enabled, it logs any statements modifications performed by the parsing operations,
+// instructing users to rewrite the statements accordingly.
+func EnableParserCollectionModifiedStatementLogging[S StatementsGetter, R any](enabled bool) ParserCollectionOption[S, R] {
+	return func(tp *ParserCollection[S, R]) error {
+		tp.modifiedStatementLogging = enabled
+		return nil
+	}
+}
+
+// ParseStatements parses the given statements into [R] using the configured context's ottl.Parser
+// and subsequently calling the ParsedStatementConverter function.
+// The statement's context is automatically inferred from the [Path.Context] values, choosing the
+// highest priority context found.
+// If no contexts are present in the statements, or if the inferred value is not supported by
+// the [ParserCollection], it returns an error.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
+func (pc *ParserCollection[S, R]) ParseStatements(statements S) (R, error) {
+	statementsValues := statements.GetStatements()
+	inferredContext, err := pc.contextInferrer.infer(statementsValues)
+	if err != nil {
+		return *new(R), err
+	}
+
+	if inferredContext == "" {
+		return *new(R), fmt.Errorf("unable to infer context from statements [%v], path's first segment must be a valid context name", statementsValues)
+	}
+
+	return pc.ParseStatementsWithContext(inferredContext, statements, false)
+}
+
+// ParseStatementsWithContext parses the given statements into [R] using the configured
+// context's ottl.Parser and subsequently calling the ParsedStatementConverter function.
+// Differently from ParseStatements, it uses the provided context and does not infer it
+// automatically. The context valuer must be supported by the [ParserCollection],
+// otherwise an error is returned.
+// If the statement's Path does not provide their Path.Context value, the prependPathsContext
+// argument should be set to true, so it rewrites the statements prepending the missing paths
+// contexts.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
+func (pc *ParserCollection[S, R]) ParseStatementsWithContext(context string, statements S, prependPathsContext bool) (R, error) {
```

Comment:
Instead of taking an generic `S` can this take []string like the parser's ParseStatements?